### ISSUE TITLE
RFC: Re-introduce weighted test Ordering with Test Load Balancing

### ISSUE THREAD

**Issue by step2yeung** [ember-cli/ember-exam#134]
This RFC outlines a plan to re-introduce weighted test ordering by test types with test load balancing in Ember Exam for Ember applications. The goals of this RFC is to:
- Enable scalable test executions 
- Optimize the speed of test runs no matter the number of tests
- Decrease test start time by loading tests on demand

[Rendered](https://gist.github.com/step2yeung/94c01cc4dcfae5af2c2d64464233a7da)

**Comment by trentmwillis** [comment]
Thanks for this very detailed proposal @step2yeung! Overall, I am very :+1: on this idea. Great enumeration of the constraints and problem space for this feature.

That said, after reading I'm not sure this proposed solution actually fixes the stated problem.

In particular, since `--load-balance` applies to a single partition (example from the RFC):

```
--split=2 --partition=1 --load-balance=20 --weighted
--split=2 --partition=2 --load-balance=20 --weighted
```

This means that the actual partitioning will still be done statically and thus the partitions will wind up being just as unbalanced as they are today.

What I would expect instead is to have a command like the original `--weighted` flag, maybe:

```
--split=2 --parallel --balanced
```

This would then allow you to balance all the modules into equally weighted partitions as proposed in the RFC.

----

That said, I can understand the desire to have different, statically defined partitions that you can run on separate CI boxes, but that is solving a different problem from what is stated here.

**Comment by step2yeung** [comment]
Thanks for the feedback @trentmwillis !

You bring up a good point, when `--load-balance` is used with `--split`, it really just balances the tests within partition 1 against 20 number of browsers. 

In the example you refer to, the concept of `partition` applies to a CI box instead of a browser. So it could be the case that partition 1 on CI box 1 finishes before partition 2 on CI box2 (or the other way around). So, it is not truly balanced when looking at all the runs as a whole.

The above is true when running on separate CI boxes. But when it comes to a single CI box, I think it is still valid:

If I am understanding correctly `--split=2 --parallel --balanced` would sort the modules by weight with `--balanced`, then `--split=2 --parallel` would split the list of modules into two partitions, which would be ran on two browsers.
The proposed solution's equivalent would be `--weighted --load-balance=2`

More thought is needed to truly balance partitions against multiple CI boxes, but I wanted to preserve this functionality of running different partitions on separate CI box for the `--load-balance` feature. Do you feel this is a good step forward?

**Comment by trentmwillis** [comment]
> I wanted to preserve this functionality of running different partitions on separate CI box for the `--load-balance` feature.

That's totally fair. I think we just need to be clear that this won't balance the partitions generated when using `--split`. I'm not sure there is actually a good solution to allow balancing across multiple different CI boxes without adding a LOT more complexity (e.g., needing to coordinate over HTTP or WebSockets between boxes).

So I'm okay with us moving forward with this approach. If folks want balanced, parallelized runs on a single box then they can use the new feature without `--split`.

---

Here is the rest of the feedback I have on the current RFC (sorry it's a bit lengthy):

**Feedback on design**

I don't think there should be both `--load-balance` and `--weighted` options. I would just have `--load-balance` and have that always use the proposed weighting behavior. I don't see much benefit to allowing the two to be used separately.

The reproduction file idea is good, but I would expect it to be used like `--load-balance=<file-path>` rather than a totally separate command (keeps the API a bit smaller).

**Other feedback**

> Ordering of the queue of test AMD modules from slowest to fastest test types (slowest first).

This is a good goal, but with the proposed implementation I have a few thoughts:

1. Shouldn't the sorting already be taken care of by virtue of being alphabetized? At least last time I checked the test AMD modules usually come out alphabetically sorted.
2. What if someone doesn't follow the naming convention? Or perhaps uses a different language for the module names?

> Stable ordering of the server side queue of test AMD modules

Not that it matters, but this should just be a concern of the build pipeline. Assuming it generates stable outputs, the inputs should be stable.

> There is a possibility that support will be required Mocha as well.

Don't worry about this. For instance, Mocha doesn't currently support randomized test runs. So you can just detect this scenario and warn the user about it (example: https://github.com/ember-cli/ember-exam/blob/31da47677c9971a5eec5e86ea5483c1c00e20494/lib/utils/tests-options-validator.js#L108-L111).

> This design execute tests modules to run in a non-deterministic order

We should definitely document this with the feature as I can see this becoming a "gotcha" for users.

> Test Module filtering

This should be done as a totally separate feature, IMO.

> Allow a test duration threshold to be set. This will help keep the test execution time free from long running test modules, by failing test modules that run longer than the threshold.

This should be done using QUnit's built-in timeout functionality. Currently that only works at the individual test level, but I would be in favor of "chunkier" levels of timeouts.

**Comment by step2yeung** [comment]
Agreed on most of the feedback. Just to touch on a few:

```
The reproduction file idea is good, but I would expect it to be used like --load-balance=<file-path> rather than a totally separate command (keeps the API a bit smaller).
```
I forgot to update that section, the command should be:
`--replay-execution=path/to/test-execution-1234.json --replay-browser=x,y`   *without the need for `--load-balance`
I thought about this, if we use `--load-balance`, the value accepted by the command now has two responsibility: path to file to replay previous runs and number to indicate the number of browser to load balance against. This seemed to overload the command. Hence I added `--replay-execution` parameter to make it explicit.

```
This is a good goal, but with the proposed implementation I have a few thoughts:

1. Shouldn't the sorting already be taken care of by virtue of being alphabetized? At least last time I checked the test AMD modules usually come out alphabetically sorted.
2. What if someone doesn't follow the naming convention? Or perhaps uses a different language for the module names?
```
1. You are correct that the module comes sorted by the path, I was calling it out here to be explicit. I don't think work needs to be done to make sure its sorted.
2. Good callout. I think to start with, we rely on default naming convention that will be called out in documentation. But we can look to make the sorting more customizable.

**Comment by step2yeung** [comment]
I would like to say this is now implemented with v2.1.4.
